File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/base/Clickable.java
package com.epam.jdi.uitests.mobile.appium.elements.base;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.base.IClickable;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;

import static java.lang.String.format;

/**
 * Created by Roman_Iovlev on 7/6/2015.
 */
public class Clickable extends Element implements IClickable {
    public Clickable() {
    }

    public Clickable(By byLocator) {
        super(byLocator);
    }

    public Clickable(WebElement webElement) {
        super(webElement);
    }

    protected void clickJSAction() {
        jsExecutor().executeScript("arguments[0].click();", getWebElement());
    }

    protected void clickAction() {
        getWebElement().click();
    }

    public final void click() {
        actions.click(this::clickAction);
    }

    public void clickByXY(int x, int y) {
        invoker.doJAction(format("Click on Element with coordinates (x,y) = (%s, %s)", x, y),
                () -> new Actions(getDriver())
                        .moveToElement(getWebElement(), x, y).click().build().perform());
    }
}

File: C#.Net/JDI/Web/Selenium/Elements/Base/Clickable.cs
using Epam.JDI.Core;
using Epam.JDI.Core.Interfaces.Base;
using OpenQA.Selenium;
using OpenQA.Selenium.Interactions;
using System;
using Epam.JDI.Web.Selenium.Base;

namespace Epam.JDI.Web.Selenium.Elements.Base
{
    public class Clickable : WebElement, IClickable
    {
        public Clickable() : this (null) { }
        public Clickable(By byLocator = null, IWebElement webElement = null) 
            : base(byLocator, webElement) { }
        
        protected Action<WebBaseElement> ClickAction = cl => cl.WebElement.Click();
 
        public void Click()
        {
            Actions.Click(ClickAction);
        }

        protected Action<WebBaseElement> ClickJsAction = cl => cl.JsExecutor.ExecuteScript("arguments[0].click();", cl.WebElement);

        public void ClickByXY(int x, int y)
        {
            Invoker.DoJAction($"Click on Element with coordinates (x,y) = ({x},{y})",
            el =>
                {
                    var element = WebElement;
                    new Actions(WebDriver).MoveToElement(WebElement, x, y).Click().Build().Perform();
                }
            );

        }
    }
}
